### ISSUE TITLE
Extensible Records?

### ISSUE THREAD

**Issue by rabipelais** [nikita-volkov/record#18]
I think it would be nice to be able to do something like this:

``` haskell
type Person = [r| {name :: String, age :: Int} |]
type HappyPerson = [r| {name :: String, age :: Int, favouriteFood :: String} |]

saruman :: Person
saruman = [r| {name = "Saruman", age = 2019} |]

happy :: Person -> HappyPerson
happy p = p <+> f
  where
    f ::  [r| {favouriteFood :: String} |]
    f = [r| {favouriteFood = "Lemon Pie"} |]
```

I think it could be done with a type class

``` haskell
class Extensible a b c where
   (<+>) :: a -> b -> c
```

which could be instantiated with Template Haskell (and that would be an ungodly amount of instances...). But perhaps there is some hairy type extension that could help?

A possible problem would be the ordering of the fields.

Do you know how the Elm people do it?

**Comment by nikita-volkov** [comment]
> A possible problem would be the ordering of the fields.

Yes, and unfortunately I just can't imagine a solution to this, which does not involve making this a special case in the compiler.

**Comment by aavogt** [comment]
With ghc >= 7.8 you have `GHC.TypeLits.CmpSymbol` to sort. https://gist.github.com/aavogt/035d8e6374e50e400035 seems to work

**Comment by nikita-volkov** [comment]
Seems very intriguing. I'll check it out later. Thanks.

**Comment by aavogt** [comment]
I've cleaned up/commented that file a bit and put it in https://github.com/aavogt/record/blob/append/library/Record/Append.hs
